tsweb/server.go

```diff
@@ -0,0 +1,385 @@
+			TLSConfig:    s.CertManager.TLSConfig(),
+		}
+		s.HTTPS.TLSConfig.MinVersion = tls.VersionTLS12
+		s.HTTP = &http.Server{
+			Addr:         ":80",
+			Handler:      s.CertManager.HTTPHandler(Port80Handler{s}),
+			ReadTimeout:  defaultTimeout,
+			WriteTimeout: defaultTimeout,
+			IdleTimeout:  defaultTimeout,
+		}
+		hstsVal := "max-age=63072000"
+		if !cfg.HSTSNoSubdomains {
+			hstsVal += "; includeSubDomains"
+		}
+		s.TLSHeaders.Set("Strict-Transport-Security", hstsVal)
+	default:
+		s.HTTP = &http.Server{
+			Addr:         cfg.Addr,
+			Handler:      s,
+			ReadTimeout:  defaultTimeout,
+			WriteTimeout: defaultTimeout,
+			IdleTimeout:  defaultTimeout,
+		}
+	}
+	s.AlwaysHeaders.Set("Content-Security-Policy", "default-src 'none'; frame-ancestors 'none'; form-action 'none'; base-uri 'self'; block-all-mixed-content; plugin-types 'none'")
+	s.AlwaysHeaders.Set("X-Frame-Options", "DENY")
+	s.AlwaysHeaders.Set("X-Content-Type-Options", "nosniff")
+
+	return s
+}
+
+// ListenAndServe listens on the TCP network addresses s.HTTP.Addr and
+// s.HTTPS.Addr (if any) and then calls Serve or ServeTLS to handle
+// incoming requests.
+//
+// If s.HTTP.Addr is blank, ":http" is used.
+// If s.HTTPS.Addr is blank, ":https" is used.
+func (s *Server) ListenAndServe() error {
+	errCh := make(chan error, 2)
+
+	if s.HTTP != nil {
+		go func() { errCh <- s.HTTP.ListenAndServe() }()
+	}
+	if s.HTTPS != nil {
+		go func() { errCh <- s.HTTPS.ListenAndServeTLS("", "") }()
+	}
+
+	err := <-errCh
+	if err == http.ErrServerClosed {
+		return nil
+	}
+	return err
+}
+
+// close immediately closes all listeners and connections, except
+// hijacked Conns. Returns any error returned from closing underlying
+// listeners.
+func (s *Server) Close() error {
+	var err error
+	if s.HTTP != nil {
+		err = s.HTTP.Close()
+	}
+	if s.HTTPS != nil {
+		err2 := s.HTTPS.Close()
+		if err == nil {
+			err = err2
+		}
+	}
+	return err
+}
+
+// ServeHTTP wraps s.Handler and adds the following features:
+//  - Initializes default values for response headers from
+//    s.AlwaysHeaders and s.TLSHeaders (if the request is over TLS).
+//  - Sends requests for /debug/* to s.Debug.
+//  - Maintains HTTP and TLS expvar metrics.
+//  - If s.RequestLog is non-nil, writes out JSON AccessLogRecord
+//    structs when requests complete.
+//  - Injects tailscale helpers into the request context, so that
+//    helpers like Err and SuppressLogging work.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer s.httpRequests.Add(1)
+	s.httpActive.Add(1)
+	defer s.httpActive.Add(-1)
+	if r.TLS != nil {
+		label := "unknown"
+		switch r.TLS.Version {
+		case tls.VersionTLS10:
+			label = "1.0"
+		case tls.VersionTLS11:
+			label = "1.1"
+		case tls.VersionTLS12:
+			label = "1.2"
+		case tls.VersionTLS13:
+			label = "1.3"
+		}
+		defer s.tlsRequests.Add(label, 1)
+		s.tlsActive.Add(label, 1)
+		defer s.tlsActive.Add(label, -1)
+	}
+
+	for k, v := range s.AlwaysHeaders {
```

Comment:
how about only set these if they're not already present in the map? Then callers can opt-out per Handler if they need to (e.g. be embeddable in a frame)

That's what Go does itself for things like the "Date" header... it's automatic to add Date unless you set the "Date" map key to `nil`.